build_resolvers/lib/src/resolver.dart

```diff
@@ -123,31 +126,102 @@ class AnalyzerResolver implements ReleasableResolver {
+
 class AnalyzerResolvers implements Resolvers {
-  final AnalyzerResolver _resolver;
-  final BuildAssetUriResolver _uriResolver;
+  /// Nullable, the default analysis options are used if not provided.
+  final AnalysisOptions _analysisOptions;
+
+  /// A function that returns the path to the SDK summary when invoked.
+  ///
+  /// Defaults to [_defaultSdkSummaryGenerator].
+  final Future<String> Function() _sdkSummaryGenerator;
+
+  /// Lazy, all access must be preceded by a call to [_ensureInitialized].
+  AnalyzerResolver _resolver;
+  BuildAssetUriResolver _uriResolver;
 
-  AnalyzerResolvers._(this._resolver, this._uriResolver);
+  /// Nullable, should not be accessed outside of [_ensureInitialized].
+  Future<void> _initialized;
+
+  AnalyzerResolvers(
+      [this._analysisOptions, Future<String> Function() sdkSummaryGenerator])
+      : _sdkSummaryGenerator =
+            sdkSummaryGenerator ?? _defaultSdkSummaryGenerator;
 
   /// Create a Resolvers backed by an [AnalysisContext] using options
-  /// [analysisOptions].
-  ///
-  /// If no argument is passed a default AnalysisOptions is used.
-  factory AnalyzerResolvers([AnalysisOptions analysisOptions]) {
-    var uriResolver = BuildAssetUriResolver();
-    var driver = analysisDriver(uriResolver, analysisOptions);
-    return AnalyzerResolvers._(
-        AnalyzerResolver(driver, uriResolver), uriResolver);
+  /// [_analysisOptions].
+  Future<void> _ensureInitialized() {
+    return _initialized ??= () async {
+      _uriResolver = BuildAssetUriResolver();
+      var driver = analysisDriver(
+          _uriResolver, _analysisOptions, await _sdkSummaryGenerator());
+      _resolver = AnalyzerResolver(driver, _uriResolver);
+    }();
   }
 
   @override
   Future<ReleasableResolver> get(BuildStep buildStep) async {
+    await _ensureInitialized();
+
     await _uriResolver.performResolve(
         buildStep, [buildStep.inputId], _resolver._driver);
     return PerActionResolver(_resolver, [buildStep.inputId]);
   }
 
   /// Must be called between each build.
   @override
-  void reset() => _uriResolver.seenAssets.clear();
+  void reset() {
+    _uriResolver?.seenAssets?.clear();
+  }
+}
+
+/// Lazily creates a summary of the users SDK and caches it under
+/// `.dart_tool/build_resolvers`.
+///
+/// This is only intended for use in typical dart packages, which must
+/// have an already existing `.dart_tool` directory (this is how we
+/// validate we are running under a typical dart package and not a custom
+/// environment).
+Future<String> _defaultSdkSummaryGenerator() async {
+  var dartToolPath = '.dart_tool';
+  if (!await Directory(dartToolPath).exists()) {
+    throw StateError(
+        'The default analyzer resolver can only be used when the current '
+        'working directory is a standard pub package.');
+  }
+
+  var cacheDir = p.join(dartToolPath, 'build_resolvers');
+  var summaryPath = p.join(cacheDir, 'sdk.sum');
+  var versionFile = File(summaryPath + '.version');
```

Comment:
[nit] `'$summaryPath.version'`

https://dart.dev/guides/language/effective-dart/usage#prefer-using-interpolation-to-compose-strings-and-values